<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Søren B. Vilsen" />


<title>Short introduction to an STRMPS workflow</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Short introduction to an STRMPS
workflow</h1>
<h4 class="author">Søren B. Vilsen</h4>
<h4 class="date">Last updated: 2025-02-22</h4>



<p>The <code>STRMPS</code> package is designed to extract and collect
the short tandem repeat (STR) information from the <code>fastq</code>
files produced by massively parallel sequencing (MPS).</p>
<div id="example-sequences" class="section level3">
<h3>Example sequences</h3>
<p>The <code>STRMPS</code>-package depends on <code>R</code> (&gt;=
4.4), <code>methods</code>, <code>utils</code>, <code>tidyr</code>,
<code>tibble</code>, <code>dplyr</code>, <code>stringr</code>,
<code>purrr</code>, <code>parallel</code>, as well as the bioconductor
packages <code>Biostrings</code>, <code>pwalign</code>,
<code>ShortRead</code>, and <code>IRanges</code>.</p>
<p>Using the <code>readFastq</code> function, we can load a data file
into <code>R</code>.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;Biostrings&quot;</span>)</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;pwalign&quot;</span>)</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;ShortRead&quot;</span>)</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a>readPath <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&#39;extdata&#39;</span>, <span class="st">&#39;sampleSequences.fastq&#39;</span>, <span class="at">package =</span> <span class="st">&#39;STRMPS&#39;</span>)</span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a>sequences <span class="ot">&lt;-</span> <span class="fu">readFastq</span>(readPath)</span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a>sequences<span class="sc">@</span>sread</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="co">#&gt; DNAStringSet object of length 20237:</span></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="co">#&gt;         width seq</span></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="co">#&gt;     [1]    70 CTGTTCTAAGTACAGTAAAGATAGATAGATAG...GGGTGAGATTTCTATCTATGAAGGCAGTTAC</span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a><span class="co">#&gt;     [2]    70 CTGTTCTAAGTACAGTAAAGATAGATAGATAG...GGGTGAGATTTCTATCTATGAAGGCAGTTAC</span></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a><span class="co">#&gt;     [3]    70 CTGTTCTAAGTACAGTAAAGATAGATAGATAG...GGGTGAGATTTCTATCTATGAAGGCAGTTAC</span></span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a><span class="co">#&gt;     [4]    70 CTGTTCTAAGTACAGTAAAGATAGATAGATAG...GGGTGAGATTTCTATCTATGAAGGCAGTTAC</span></span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a><span class="co">#&gt;     [5]    70 CTGTTCTAAGTACAGTAAAGATAGATAGATAG...GGGTGAGATTTCTATCTATGAAGGCAGTTAC</span></span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a><span class="co">#&gt;     ...   ... ...</span></span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a><span class="co">#&gt; [20233]   144 GGACAGATGATACCGCATGGTCCGCTAAAGCT...TACAATTCGGACCAAGTCACATACTGATTAT</span></span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a><span class="co">#&gt; [20234]   144 GGACAGATGATACCGCATGGTCCGCTAAAGCT...TACAATTCGGACCAAGTCACATACTGATTAT</span></span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a><span class="co">#&gt; [20235]   144 GGACAGATGATACCGCATGGTCCGCTAAAGCT...TACAATTCGGACCAAGTCACATACTGATTAT</span></span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a><span class="co">#&gt; [20236]   144 GGACAGATGATACCGCATGGTCCGCTAAAGCT...TACAATTCGGACCAAGTCACATACTGATTAT</span></span>
<span id="cb2-14"><a href="#cb2-14" tabindex="-1"></a><span class="co">#&gt; [20237]   144 GGACAGATGATACCGCATGGTCCGCTAAAGCT...TACAATTCGGACCAAGTCACATACTGATTAT</span></span></code></pre></div>
</div>
<div id="flanking-regions" class="section level3">
<h3>Flanking regions</h3>
<p>We are interested in extracting the STR regions of these sequences.
They are extracted by searching for marker specific sequences in the
regions surrounding the STR region – called the flanking regions. The
following is an example of a <code>tibble</code> containing the
necessary flanking region information. NOTE: this is just an example,
the following loaded flanking region <code>tibble</code> is not usable
for actual applications.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;STRMPS&quot;</span>)</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;flankingRegions&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">head</span>(flankingRegions, <span class="dv">5</span>)</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="co">#&gt; # A tibble: 5 × 10</span></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="co">#&gt;   Marker     Chromosome Type  ForwardFlank ReverseFlank Motif MotifLength Offset</span></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="co">#&gt;   &lt;chr&gt;      &lt;chr&gt;      &lt;chr&gt; &lt;chr&gt;        &lt;chr&gt;        &lt;chr&gt;       &lt;int&gt;  &lt;int&gt;</span></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a><span class="co">#&gt; 1 AMELOGENIN X/Y        AUTO… CCCTGGGCTCT… CAGCTTCCCAGT TCAA…           6     63</span></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a><span class="co">#&gt; 2 CSF1PO     5          AUTO… CTGTTCTAAGTA CTATCTATGAA… ATCT…           4     20</span></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a><span class="co">#&gt; 3 D10S1248   10         AUTO… GTCACAAACAT… GTTCCTTTAATA GGAA…           4     41</span></span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a><span class="co">#&gt; 4 D12S391    12         AUTO… CTGTATTAGTA… GCTGGAGACC   GTCT…           4    112</span></span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a><span class="co">#&gt; 5 D13S317    13         AUTO… GACTCTCTGGAC CTGCCTATGGC… TATC…           4     78</span></span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a><span class="co">#&gt; # ℹ 2 more variables: ForwardShift &lt;dbl&gt;, ReverseShift &lt;dbl&gt;</span></span></code></pre></div>
<p>The columns of the <code>tibble</code> contains the following
information:</p>
<ul>
<li><code>Marker</code> and <code>Type</code> are the name and
chromosomal type (autosomal, X, or Y) of each marker.</li>
<li><code>ForwardFlank</code> and <code>ReverseFlank</code> contain the
marker specific sequences used to identify the STR regions. The forward
and reverse flanks should occur before and after the STR regions,
respectively.</li>
<li><code>Motif</code> and <code>MotifLength</code> are the structure
and length of the STR regions motif, respectively.</li>
<li><code>Offset</code> is used to adjust the numeric allele designation
so it can be compared to the corresponding CE allele designation.</li>
<li><code>ForwardShift</code> and <code>ReverseShift</code> (not shown)
are used to trim the extracted sequences to just include the STR
regions. If these are set to zero, then the extracted regions contain
some useful flanking region information.</li>
</ul>
</div>
<div id="identification" class="section level3">
<h3>Identification</h3>
<p>Using the <code>flankingRegions</code> file, we can identify the STR
regions of the sequences by calling the <code>identifySTRRegions</code>
function. The function takes four arguments:</p>
<ul>
<li><code>reads</code>: The sequences to search through.</li>
<li><code>flankingRegions</code>: A <code>tibble</code> or
<code>data.frame</code> in the same style as shown above.</li>
<li><code>numberOfMutation</code>: The number of allowed mismatches,
when searching the sequences.</li>
<li><code>control</code>: A control object setting additional parameters
(to see more type <code>?identifySTRRegions.control</code>).</li>
</ul>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a>identifiedSTRs <span class="ot">&lt;-</span> <span class="fu">identifySTRRegions</span>(</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>    <span class="at">reads =</span> sequences,</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>    <span class="at">flankingRegions =</span> flankingRegions,</span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a>    <span class="at">numberOfMutation =</span> <span class="dv">1</span>,</span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a>    <span class="at">control =</span> <span class="fu">identifySTRRegions.control</span>(</span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a>        <span class="at">numberOfThreads =</span> <span class="dv">1</span>,</span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a>        <span class="at">includeReverseComplement =</span> <span class="cn">FALSE</span></span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a>    )</span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a>)</span></code></pre></div>
<p>The function returns a list with the following:</p>
<ul>
<li><code>n_reads</code>: The total number of reads in the supplied
<code>reads</code> object.</li>
<li><code>reverseComplement</code>: TRUE/FALSE value – did we search for
the reverse complementary flanking regions?</li>
<li><code>identifiedMarkers</code>: A list containing indecies, and the
start and end positions of STR regions for each of the markers in the
<code>flankingRegiions</code> object.</li>
<li><code>identifiedMarkersSequencesUniquelyAssigned</code>: Almost
identical to <code>identifiedMarkers</code>, but if multiple markers
have been found in a sequences it is disregarded.</li>
<li><code>remainingSequences</code>: A vector of indecies of sequences
where no marker was identified.</li>
</ul>
<p>An example of what is stored in the
<code>identifiedMarkersSequencesUniquelyAssigned</code> list</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">names</span>(identifiedSTRs<span class="sc">$</span>identifiedMarkersSequencesUniquelyAssigned<span class="sc">$</span>CSF1PO)</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="co">#&gt;  [1] &quot;name&quot;                          &quot;matchedSeq&quot;                   </span></span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="co">#&gt;  [3] &quot;startForward&quot;                  &quot;endForward&quot;                   </span></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a><span class="co">#&gt;  [5] &quot;startReverse&quot;                  &quot;endReverse&quot;                   </span></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a><span class="co">#&gt;  [7] &quot;trimmedIncludingFlanks&quot;        &quot;trimmedQualityIncludingFlanks&quot;</span></span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a><span class="co">#&gt;  [9] &quot;trimmed&quot;                       &quot;trimmedQuality&quot;</span></span></code></pre></div>
<p>That is, we find the name of the marker, the indecies of the
sequences identified as belonging to the marker, the start of the
forward flank, the end of the reverse flank, and then trimmed versions
of the sequences, where ‘junk’ has been removed (any information outside
the searched for flanking sequences).</p>
</div>
<div id="aggregation" class="section level3">
<h3>Aggregation</h3>
<p>Given the identified STR regions, we can aggregate the identified
strings, to get their coverage. When doing so, we also supply the length
of the motifs of every marker, and their type, i.e. whether the marker
is located on an autosomal, the X, or the Y chromosome. In order to
supply the motif length and type information, we need to know which
markers were actually identified, and in which order they were
identified.</p>
<p>This could be done as follows:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a>sortedIncludedMarkers <span class="ot">&lt;-</span> <span class="fu">sapply</span>(</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>    <span class="fu">names</span>(identifiedSTRs<span class="sc">$</span>identifiedMarkersSequencesUniquelyAssigned),</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a>    <span class="cf">function</span>(m) <span class="fu">which</span>(m <span class="sc">==</span> flankingRegions<span class="sc">$</span>Marker)</span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a>)</span></code></pre></div>
<p>The aggregation is then performed by calling the
<code>stringCoverage</code> function with the arguments
<code>extractedReadsListObject</code>, <code>motifLength</code>,
<code>Type</code>, and <code>control</code>. In the <code>control</code>
argument, we pass an argument called <code>simpleReturn</code>. If this
argument is set to <code>FALSE</code>, the aggregation is performed
based on both the STR region <strong>and</strong> the forward and
reverse flanks. If it is set to <code>TRUE</code>, the aggregation is
just performed based on the STR regions. That is, if it is
<code>FALSE</code>, we will find more distinct strings, with lower
coverage. Note that when setting <code>simpleReturn = TRUE</code>, the
quality is also summarised by the geometric average for each base.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a>stringCoverageList <span class="ot">&lt;-</span> <span class="fu">stringCoverage</span>(</span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>    <span class="at">extractedReadsListObject =</span> identifiedSTRs,</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>    <span class="at">flankingRegions =</span> flankingRegions,</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a>    <span class="at">control =</span> <span class="fu">stringCoverage.control</span>(</span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a>        <span class="at">numberOfThreads =</span> <span class="dv">1</span>,</span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a>        <span class="at">trace =</span> <span class="cn">FALSE</span>,</span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a>        <span class="at">simpleReturn =</span> <span class="cn">TRUE</span></span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a>    )</span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a>)</span></code></pre></div>
<p>The <code>stringCoverage</code> returns a list of
<code>tibbles</code> one for each of the identified markers. The result
can be seen here:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a>stringCoverageList<span class="sc">$</span>CSF1PO</span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a><span class="co">#&gt; # A tibble: 3 × 9</span></span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a><span class="co">#&gt;   Marker BasePairs Allele Type      MotifLength Region Coverage AggregateQuality</span></span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a><span class="co">#&gt;   &lt;chr&gt;      &lt;int&gt;  &lt;dbl&gt; &lt;chr&gt;           &lt;int&gt; &lt;chr&gt;     &lt;int&gt; &lt;chr&gt;           </span></span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a><span class="co">#&gt; 1 CSF1PO        38    9.5 AUTOSOMAL           4 CAGTA…      955 [[[[[[[[[[[[[[[…</span></span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a><span class="co">#&gt; 2 CSF1PO        50   12.5 AUTOSOMAL           4 CAGTA…       37 [[[[[[[[[[[[[[[…</span></span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a><span class="co">#&gt; 3 CSF1PO        54   13.5 AUTOSOMAL           4 CAGTA…      994 [[[[[[[[[[[[[[[…</span></span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a><span class="co">#&gt; # ℹ 1 more variable: Quality &lt;list&gt;</span></span></code></pre></div>
<p>We see on marker <code>CSF1PO</code>, we have found three distinct
markers two with a coverage above 950, and one with a coverage of 37. If
the sample only contained a single contributor, we would expect that the
two strings with a coverage of above 950 to be the alleles of the
individual. While the string with low coverage is either a stutter (as
it is in the stutter position), or an error.</p>
</div>
<div id="genotype-identification" class="section level3">
<h3>Genotype identification</h3>
<p>If the aggregated <code>stringCoverageList</code>-object, contains
DNA from a single contributor <strong>and</strong> was made with a large
amount of input DNA, we can determine the genotype of the contributor. A
way of doing so is to identify the string with the largest coverage, and
determine whether or not another allele is within some pre-defined
heterozygous threshold (by default this is set to <span class="math inline">\(0.35\)</span>).</p>
<p>We determine the genotype of each marker, by using the
<code>getGenotype</code> function.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a>genotypeList <span class="ot">&lt;-</span> <span class="fu">getGenotype</span>(stringCoverageList)</span></code></pre></div>
<p>The created <code>genotypeList</code>-object contains the genotypes
determined by the <code>thresholdHeterozygosity</code>. Continuing the
example from before, the genotypes of the marker <code>CSF1PO</code>
are:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a>genotypeList<span class="sc">$</span>CSF1PO</span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a><span class="co">#&gt; # A tibble: 2 × 10</span></span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a><span class="co">#&gt;   Marker BasePairs Allele Type      MotifLength Region Coverage AggregateQuality</span></span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a><span class="co">#&gt;   &lt;chr&gt;      &lt;int&gt;  &lt;dbl&gt; &lt;chr&gt;           &lt;int&gt; &lt;chr&gt;     &lt;int&gt; &lt;chr&gt;           </span></span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a><span class="co">#&gt; 1 CSF1PO        54   13.5 AUTOSOMAL           4 CAGTA…      994 [[[[[[[[[[[[[[[…</span></span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a><span class="co">#&gt; 2 CSF1PO        38    9.5 AUTOSOMAL           4 CAGTA…      955 [[[[[[[[[[[[[[[…</span></span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a><span class="co">#&gt; # ℹ 2 more variables: Quality &lt;list&gt;, Indices &lt;int&gt;</span></span></code></pre></div>
<p>We see that the <code>getGenotype</code> function has identified the
two strings with coverage above 950, as the potential alleles of the
marker for the contributor of the sample. Furthermore, note that the
function adds a column called <code>Indices</code> containing the
indices of the corresponding strings in the
<code>stringCoverageList</code>-object. That is, in this example,
<code>Indices</code> contains the values 1 and 3.</p>
</div>
<div id="noise-identification" class="section level3">
<h3>Noise identification</h3>
<p>In a similar way to the genotype identification, we implemented
simple rules for noise identification. A string is classified as noise
if it is less than the coverage of the most prevalent string times
<code>thresholdSignal</code>. By default this factor is set to <span class="math inline">\(1\%\)</span>.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a>noiseList <span class="ot">&lt;-</span> <span class="fu">identifyNoise</span>(stringCoverageList, <span class="at">thresholdSignal =</span> <span class="fl">0.03</span>)</span></code></pre></div>
<p>The <code>identifyNoise</code> function returns a list with an
element for every marker in the <code>stringCoverageList</code>-object.
If an observation is classified as noise, it is then removed from the
data. Returning to our example from before, marker <code>CSF1PO</code>
has no strings with less than <span class="math inline">\(1\%\)</span>
of the coverage of the most prevalent string. That is, the following
still contains three distinct sequences:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a>noiseList<span class="sc">$</span>CSF1PO</span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a><span class="co">#&gt; # A tibble: 2 × 10</span></span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a><span class="co">#&gt;   Marker BasePairs Allele Type      MotifLength Region Coverage AggregateQuality</span></span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a><span class="co">#&gt;   &lt;chr&gt;      &lt;int&gt;  &lt;dbl&gt; &lt;chr&gt;           &lt;int&gt; &lt;chr&gt;     &lt;int&gt; &lt;chr&gt;           </span></span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a><span class="co">#&gt; 1 CSF1PO        54   13.5 AUTOSOMAL           4 CAGTA…      994 [[[[[[[[[[[[[[[…</span></span>
<span id="cb13-6"><a href="#cb13-6" tabindex="-1"></a><span class="co">#&gt; 2 CSF1PO        38    9.5 AUTOSOMAL           4 CAGTA…      955 [[[[[[[[[[[[[[[…</span></span>
<span id="cb13-7"><a href="#cb13-7" tabindex="-1"></a><span class="co">#&gt; # ℹ 2 more variables: Quality &lt;list&gt;, Indices &lt;int&gt;</span></span></code></pre></div>
<p>Note that as with the <code>getGenotype</code> function, the
<code>identifyNoise</code> function adds a column called
<code>Indices</code> containing the indices of the corresponding strings
in the <code>stringCoverageList</code>-object. That is, in this example,
<code>Indices</code> contains the values 1, 2, and 3.</p>
</div>
<div id="stuttering" class="section level3">
<h3>Stuttering</h3>
<p>Stuttering refers to the loss (or gain) of a motif in the allele
sequence during PCR amplification. A predictor of the rate of stuttering
(quantified by the stutter ratio or stutter proportion), is the Block
Length of the Missing Motif (BLMM). If we compare the allele to a
stutter sequence, then we can identify the approximate location (down to
the nearest block) of the motif which stuttered. The BLMM is the length
of the block (sub-sequence), which has lost a motif.</p>
<p>The <code>findStutter</code> function for every called allele, found
with the <code>getGenotype</code> function, finds all potential stutters
of the alleles and calculates the stutter ratio and stutter proportions
between the allele and potential stutter. It also identifies the missing
motif and finds the length of the corresponding block, as well as the
longest uninterrupted stretch (LUS), i.e. the length of the longest
block. It takes a merge of the <code>stringCoverageList</code> and the
<code>genotypeList</code> created using the
<code>mergeGenotypeStringCoverage</code> function.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a>stringCoverageGenotypeList <span class="ot">&lt;-</span> <span class="fu">mergeGenotypeStringCoverage</span>(stringCoverageList, genotypeList)</span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a></span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a>stutterList <span class="ot">&lt;-</span> <span class="fu">findStutter</span>(stringCoverageGenotypeList)</span>
<span id="cb14-4"><a href="#cb14-4" tabindex="-1"></a>stutterTibble <span class="ot">&lt;-</span> <span class="fu">subset</span>(<span class="fu">do.call</span>(<span class="st">&quot;rbind&quot;</span>, stutterList), <span class="sc">!</span><span class="fu">is.na</span>(Genotype))</span>
<span id="cb14-5"><a href="#cb14-5" tabindex="-1"></a><span class="fu">head</span>(stutterTibble, <span class="dv">5</span>)</span>
<span id="cb14-6"><a href="#cb14-6" tabindex="-1"></a><span class="co">#&gt; # A tibble: 5 × 20</span></span>
<span id="cb14-7"><a href="#cb14-7" tabindex="-1"></a><span class="co">#&gt;   Marker  Genotype    ParentAllele ParentString        ParentLUS ParentLUSLength</span></span>
<span id="cb14-8"><a href="#cb14-8" tabindex="-1"></a><span class="co">#&gt;   &lt;chr&gt;   &lt;chr&gt;              &lt;dbl&gt; &lt;chr&gt;               &lt;chr&gt;               &lt;int&gt;</span></span>
<span id="cb14-9"><a href="#cb14-9" tabindex="-1"></a><span class="co">#&gt; 1 CSF1PO  9.5,13.5            13.5 CAGTAAAGATAGATAGAT… [AGAT]9                 9</span></span>
<span id="cb14-10"><a href="#cb14-10" tabindex="-1"></a><span class="co">#&gt; 2 D13S317 29.5                29.5 AATGGCGGGTCGATGAGG… [TATC]10               10</span></span>
<span id="cb14-11"><a href="#cb14-11" tabindex="-1"></a><span class="co">#&gt; 3 D18S51  31.75,37.75         37.8 GCTGAAGAAAGAAAGAAA… [AAGA]22               22</span></span>
<span id="cb14-12"><a href="#cb14-12" tabindex="-1"></a><span class="co">#&gt; 4 D3S1358 31.5,32.5           32.5 ATTACTCCATTTGGCCAA… [TCTA]12               12</span></span>
<span id="cb14-13"><a href="#cb14-13" tabindex="-1"></a><span class="co">#&gt; 5 D5S818  18                  18   CAGAGTAGAGTTGAGATA… [AGAT]13               13</span></span>
<span id="cb14-14"><a href="#cb14-14" tabindex="-1"></a><span class="co">#&gt; # ℹ 14 more variables: ParentCoverage &lt;int&gt;, NeighbourAllele &lt;dbl&gt;,</span></span>
<span id="cb14-15"><a href="#cb14-15" tabindex="-1"></a><span class="co">#&gt; #   NeighbourString &lt;chr&gt;, Block &lt;chr&gt;, MissingMotif &lt;chr&gt;,</span></span>
<span id="cb14-16"><a href="#cb14-16" tabindex="-1"></a><span class="co">#&gt; #   BlockLengthMissingMotif &lt;int&gt;, NeighbourCoverage &lt;int&gt;,</span></span>
<span id="cb14-17"><a href="#cb14-17" tabindex="-1"></a><span class="co">#&gt; #   NeighbourRatio &lt;dbl&gt;, NeighbourProportion &lt;dbl&gt;,</span></span>
<span id="cb14-18"><a href="#cb14-18" tabindex="-1"></a><span class="co">#&gt; #   FLAGStutterIdentifiedMoreThanOnce &lt;lgl&gt;, FLAGMoreThanTwoAlleles &lt;lgl&gt;,</span></span>
<span id="cb14-19"><a href="#cb14-19" tabindex="-1"></a><span class="co">#&gt; #   FLAGAlleleDifferenceOne &lt;lgl&gt;, FLAGMoreThanOneBlock &lt;lgl&gt;,</span></span>
<span id="cb14-20"><a href="#cb14-20" tabindex="-1"></a><span class="co">#&gt; #   FLAGBlocksWithDifferentLengths &lt;lgl&gt;</span></span></code></pre></div>
<p>Note that the output also contains a series of flags which could be
useful:</p>
<ul>
<li><code>FLAGStutterIdentifiedMoreThanOnce</code>: Is the stutter
string been identified as a stutter for more than one allele?</li>
<li><code>FLAGMoreThanTwoAlleles</code>: Were more than two alleles
provided by the merged <code>stringCoverageGenotypeList</code>
object?</li>
<li><code>FLAGAlleleDifferenceOne</code>: Was the difference between the
two alleles one motif? If <code>TRUE</code> the shorter will appear as a
stutter of the longer and, thereby, skew the results of any analysis.
These cases should be removed.<br />
</li>
<li><code>FLAGMoreThanOneBlock</code>: Was more than one block
identified when comparing the potenetial stutter and allele
sequences?</li>
<li><code>FLAGBlocksWithDifferentLengths</code>: If
<code>FLAGMoreThanOneBlock</code> was <code>TRUE</code>, did the
identified blocks have different lengths? Note: This should not be
possible in theory, but if it happens it is at most a difference of one
and the longest is always chosen (as it should be more likely).</li>
</ul>
</div>
<div id="workflow-function" class="section level3">
<h3>Workflow function</h3>
<p>Instead of calling all the above functions, we can call the workflow
function <code>STRMPSWorkflow</code>. The function takes a path to a
file, and either returns a series of <code>.RData</code> files in the
provided <code>output</code> directory, or it returns the
<code>stringCoverageList</code> object created by the
<code>stringCoverage</code> function. The function can also continue
from previously created files using the <code>continueCheckpoint</code>
argument (if the files are placed in the <code>output</code>
directory).</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="fu">STRMPSWorkflow</span>(</span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a>    read_path,</span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a>    <span class="at">control =</span> <span class="fu">workflow.control</span>(</span>
<span id="cb15-4"><a href="#cb15-4" tabindex="-1"></a>        <span class="at">restrictType =</span> <span class="st">&quot;Autosomal&quot;</span>,</span>
<span id="cb15-5"><a href="#cb15-5" tabindex="-1"></a>        <span class="at">numberOfThreads =</span> <span class="dv">1</span></span>
<span id="cb15-6"><a href="#cb15-6" tabindex="-1"></a>    )</span>
<span id="cb15-7"><a href="#cb15-7" tabindex="-1"></a>)</span></code></pre></div>
<p>There also exists a batch version of the function called
<code>STRMPSWorkflowBatch</code> which takes an <code>input</code>
directory containing the <code>.fastq</code> files to be analysed.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
